README
This program (or rather module) gives the ability to load and parse a TTF font with the `fnt_initFont` function,
and then then use it to draw text to an SDL2 SDL_Renderer using `fnt_drawText`.
See the fonts.h header file for more detailed explanations.

This mini-subproject was originally supposed to just be a small part of my 'cgd' project, added in a single commit, however it proved to be so difficult to implement that I created a separate repo for it to make it easier to work on.


NEWS
Added a .gitignore
Made the displayed text more... appropriate
Made the fonts load from a local direcotry, rather than from somewhere in the system


test.sh
#!/bin/sh
min=1
max=3
for i in $(seq $min $max); do
    stream=/dev/null
    [ $i -eq 1 ] && stream=/dev/stdout
    bin/main $i >$stream
    echo $i/$max;
done


src/main.c
#include "fonts.h"
#include <cgd/user-input/keyboard.h>
#include <SDL2/SDL_blendmode.h>
#include <SDL2/SDL_error.h>
#include <SDL2/SDL_events.h>
#include <SDL2/SDL_pixels.h>
#include <SDL2/SDL_render.h>
#include <SDL2/SDL_stdinc.h>
#include <SDL2/SDL_surface.h>
#include <stdarg.h>
#include <stdbool.h>
#include <assert.h>
#include <stdio.h>
#include <string.h>

#define WINDOW_TITLE    "FreeType 2 SDL"
#define WINDOW_X        SDL_WINDOWPOS_CENTERED
#define WINDOW_Y        SDL_WINDOWPOS_CENTERED
#define WINDOW_WIDTH    600
#define WINDOW_HEIGHT   600
#define WINDOW_FLAGS    SDL_WINDOW_OPENGL

#define RENDERER_FLAGS  (SDL_RENDERER_ACCELERATED|SDL_RENDERER_PRESENTVSYNC|SDL_RENDERER_TARGETTEXTURE)

#define KEYBOARD_N_KEYS 7

#define LINE_HEIGHT     100.f
#define CHAR_W          50.f

#define TEXT1           "Test\t*\n12\t4,+_"
#define TEXT1_X         0
#define TEXT1_Y         0
#define TEXT2           "TES\tT"
#define TEXT2_X         200
#define TEXT2_Y         330
#define TEXT3           "1234567890"
#define TEXT3_X         0
#define TEXT3_Y         (WINDOW_HEIGHT - LINE_HEIGHT - 1)

#define MOVEMENT_SPEED  3
#define RESIZE_SPEED    1

#define SOURCE_CODE_PRO "assets/SourceCodePro-Regular.otf"
#define LIBERATION_MONO "assets/LiberationMono-Regular.ttf" 
#define FREE_SANS       "assets/FreeSansBold.otf"

#define M_SQRT2	1.41421356237309504880	/* sqrt(2) */
char FONT_PATH[FNT_TEXT_BUFFER_SIZE];

kb_Keyboard *keyboard;

int time = 0;
bool running = true;
bool displayTextRects = true, displayCharRects = true, displayGlyphRects = true;
fnt_Vector2D textOffset = { .x = 0, .y = 0 };

int max(int a, int b);

int main(int argc, char **argv)
{
    printf("Keys 1, 2 and 3 will toggle the displaying of their respective 'hitbox' outlines,\n"
        "W, S, A, D will move the whole thing,\n"
        "and the arrow keys will change the font sizes.\n");

    strncpy(FONT_PATH, LIBERATION_MONO, FNT_TEXT_BUFFER_SIZE - 1);
    if(argc == 2){
        switch(argv[1][0]){
            default: case '1':
            	break;
            case '2':
                strncpy(FONT_PATH, SOURCE_CODE_PRO, FNT_TEXT_BUFFER_SIZE - 1);
            	break;
            case '3':
                strncpy(FONT_PATH, FREE_SANS, FNT_TEXT_BUFFER_SIZE - 1);
            	break;
        }
    }
    FONT_PATH[FNT_TEXT_BUFFER_SIZE - 1] = '\0';

    SDL_Init(SDL_INIT_VIDEO);
    SDL_Window *window = SDL_CreateWindow(
        WINDOW_TITLE, WINDOW_X, WINDOW_Y, WINDOW_WIDTH, WINDOW_HEIGHT, WINDOW_FLAGS
    );
    SDL_Renderer *renderer = SDL_CreateRenderer(window, 0, RENDERER_FLAGS);
    SDL_SetRenderDrawBlendMode(renderer, SDL_BLENDMODE_BLEND);

    fnt_Font *font = fnt_initFont(FONT_PATH, renderer, CHAR_W, LINE_HEIGHT, FNT_CHARSET_ASCII, 
            FNT_FLAG_DISPLAY_GLYPH_RECTS | FNT_FLAG_DISPLAY_CHAR_RECTS | FNT_FLAG_DISPLAT_TEXT_RECTS);
    assert(font != NULL);

    keyboard = kb_initKeyboard();
    assert(keyboard != NULL);

    while(running){

        SDL_Event event;

        kb_updateKeyboard(keyboard);
        while(SDL_PollEvent(&event)){
            if(event.type == SDL_QUIT){
                running = 0;
            }
        }
        if(kb_getKey(keyboard, KB_KEYCODE_W)->pressed)
            textOffset.y -= MOVEMENT_SPEED;
        if(kb_getKey(keyboard, KB_KEYCODE_S)->pressed)
            textOffset.y += MOVEMENT_SPEED;
        if(kb_getKey(keyboard, KB_KEYCODE_A)->pressed)
            textOffset.x -= MOVEMENT_SPEED;
        if(kb_getKey(keyboard, KB_KEYCODE_D)->pressed)
            textOffset.x += MOVEMENT_SPEED;

        if(kb_getKey(keyboard, KB_KEYCODE_ARROWUP)->pressed)
            font->lineHeight -= RESIZE_SPEED * (LINE_HEIGHT / CHAR_W);
        if(kb_getKey(keyboard, KB_KEYCODE_ARROWDOWN)->pressed)
            font->lineHeight += RESIZE_SPEED * (LINE_HEIGHT / CHAR_W);
        if(kb_getKey(keyboard, KB_KEYCODE_ARROWLEFT)->pressed)
            font->charW -= RESIZE_SPEED;
        if(kb_getKey(keyboard, KB_KEYCODE_ARROWRIGHT)->pressed)
            font->charW += RESIZE_SPEED;

        if(kb_getKey(keyboard, KB_KEYCODE_DIGIT1)->down)
            font->flags ^= FNT_FLAG_DISPLAY_GLYPH_RECTS;

        if(kb_getKey(keyboard, KB_KEYCODE_DIGIT2)->down)
            font->flags ^= FNT_FLAG_DISPLAY_CHAR_RECTS;

        if(kb_getKey(keyboard, KB_KEYCODE_DIGIT3)->down)
            font->flags ^= FNT_FLAG_DISPLAT_TEXT_RECTS;

        SDL_SetRenderDrawColor(renderer, 0, 0, 0, 255);
        SDL_RenderClear(renderer);

        fnt_Vector2D textPos;
        fnt_setTextColor(font, 128, 128, time/25 % 255, 255);
        textPos = (fnt_Vector2D){ .x = TEXT1_X + textOffset.x, .y = TEXT1_Y + textOffset.y };
        fnt_drawText(font, renderer, &textPos, TEXT1);

        fnt_setTextColor(font, (int)((time + 100)*0.8) % 255, time % 100, time %255, time %10 + 200);
        textPos = (fnt_Vector2D){ .x = TEXT2_X + textOffset.x, .y = TEXT2_Y + textOffset.y };
        fnt_drawText(font, renderer, &textPos, TEXT2);

        fnt_setTextColor(font, 255, 255, 255, 128);
        textPos = (fnt_Vector2D){ .x = TEXT3_X + textOffset.x, .y = TEXT3_Y + textOffset.y };
        fnt_drawText(font, renderer, &textPos, TEXT3);

        SDL_RenderPresent(renderer);
        time++;
        SDL_Delay(16);
    }

    fnt_destroyFont(font);
    kb_destroyKeyboard(keyboard);

    SDL_DestroyRenderer(renderer);
    SDL_DestroyWindow(window);
    SDL_Quit();
    return EXIT_SUCCESS;
}


src/fonts.h
#ifndef MAIN_H
#define MAIN_H

#include <SDL2/SDL.h>
#include <SDL2/SDL_pixels.h>
#include <SDL2/SDL_rect.h>
#include <SDL2/SDL_render.h>
#include <SDL2/SDL_timer.h>
#include <SDL2/SDL_video.h>
#include <math.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <ft2build.h>
#include <sys/types.h>
#include FT_FREETYPE_H

#define FNT_ASCII_FIRST_VISIBLE_CHAR        (((int)' ') + 1)
#define FNT_ASCII_LAST_VISIBLE_CHAR         127
#define FNT_ASCII_TOTAL_VISIBLE_CHARS       (FNT_ASCII_LAST_VISIBLE_CHAR - FNT_ASCII_FIRST_VISIBLE_CHAR + 1)

#define FNT_DEFAULT_TAB_WIDTH       4
#define FNT_DEFAULT_TEXT_COLOR      (SDL_Color){ 0, 0, 0, 255 }
#define FNT_TEXT_BUFFER_SIZE        1024

typedef uint64_t fnt_uInt64;
typedef uint32_t fnt_uInt32;
typedef uint16_t fnt_uInt16;
typedef uint8_t fnt_uInt8;
typedef float_t fnt_float;
typedef double_t fnt_double;

typedef enum {
    FNT_CHARSET_ASCII,
    FNT_CHARSET_UTF8,
} fnt_Charset;

typedef struct {
    fnt_float x, y;
} fnt_Vector2D;

typedef enum {
    FNT_FLAG_DISPLAY_CHAR_RECTS     = 1,
    FNT_FLAG_DISPLAY_GLYPH_RECTS    = 2 << 0,
    FNT_FLAG_DISPLAT_TEXT_RECTS     = 2 << 1,
} fnt_Flag;

typedef struct {
    SDL_Rect srcRect;
    fnt_float offsetX, offsetY;
    fnt_float scaleX, scaleY;
} fnt_GlyphData;

typedef struct {
    /* Used internally, not modifiable */
    SDL_Texture *texture;

    fnt_GlyphData *glyphs;
    struct {
        fnt_uInt32 first;
        fnt_uInt32 last;
        fnt_uInt32 total;
    } visibleChars;
    fnt_Charset charset;

    /* User-modifiable at runtime */
    fnt_float lineHeight;
    fnt_float charW;
    fnt_uInt16 tabWidth;

    fnt_uInt16 flags;
} fnt_Font;

fnt_Font *fnt_initFont(const char *filePath, SDL_Renderer *renderer, fnt_float charW, fnt_float charH, 
        fnt_Charset charset, fnt_uInt16 flags);

void fnt_drawText(fnt_Font *fnt, SDL_Renderer *renderer, fnt_Vector2D *pos, const char *fmt, ...);

void fnt_destroyFont(fnt_Font *fnt);

/* Sets the font's texture color- and blend mode */
void fnt_setTextColor(fnt_Font *fnt, fnt_uInt8 r, fnt_uInt8 g, fnt_uInt8 b, fnt_uInt8 a);

#endif


compile-cgd-userinput-lib.sh
#!/bin/sh
_start_time="$(date +%s.%N)"

error(){
    printf "\033[31;1mERROR\033[0;0m: $1\n"
    exit $2
}
FT2SDL_DIR="$(pwd)"
CGD_DIR="$1"
LIB_NAME='libcgdengine_user-input.a'
OK_MSG='\033[32;1mOK\033[0;0m\n'

[ -d "$1" ] || {
    echo "usage: ./compile-cgd-userinput-lib.sh <cgd dir> [-v|--verbose]" >&2
    exit 1
}

OUT_STREAM=/dev/null
ERR_STREAM=/dev/null
[ "$2" = '-v' ] || [ "$2" = '--verbose' ] && {
    VERBOSE=1
	OUT_STREAM=/dev/stdout
	ERR_STREAM=/dev/stderr
}

# Copy over header files
echo -n ">> Updating header files... "
[ $VERBOSE ] && echo

headers="$(find "$CGD_DIR/user-input/" -name "*.h" | xargs echo)"

rm -vfr   "$FT2SDL_DIR/include/cgd/user-input" >$OUT_STREAM 2>$ERR_STREAM
mkdir -vp "$FT2SDL_DIR/include/cgd/user-input" >$OUT_STREAM 2>$ERR_STREAM &&
cp -vf     $headers -t "$FT2SDL_DIR/include/cgd/user-input" >$OUT_STREAM 2>$ERR_STREAM || 
    error "Failed to copy over the required header files." 2
printf $OK_MSG

# Remove and recompile the needed obj files
echo -n ">> Recompiling object files...   "
[ $VERBOSE ] && echo
object_files="$(echo "$headers" | sed -e "s/user-input\//obj\//g" -e "s/\.h/.c.o/g")"
rm -vf $object_files >$OUT_STREAM
make compile -C "$CGD_DIR" -j$(nproc --all) >$OUT_STREAM 2>$ERR_STREAM
    #|| error "Compilation failed." 3

printf $OK_MSG

# Create the .a archive
echo -n ">> Archiving... "
[ $VERBOSE ] && printf "\nAR 	%-30s %-30s\n" "lib/$LIB_NAME" "<= $object_files"
ar $VERBOSE_FLAGS rcs "$FT2SDL_DIR/lib/$LIB_NAME" $object_files >$OUT_STREAM ||
    error "Failed to create the library archive." 4

printf $OK_MSG

_elapsed_time="$(echo "$(date +%s.%N) - $_start_time" | bc)"
echo ">> Done, took $(printf "%.4f %s" "$_elapsed_time" "seconds")"
exit 0


src/fonts.c
#include "fonts.h"
#include "freetype/freetype.h"
#include <SDL2/SDL_render.h>
#include <stdlib.h>

int max(int a, int b)
{
    if(a > b)
        return a;
    else
        return b;
};
int min(int a, int b)
{
    if(a < b)
        return a;
    else
        return b;
};
double fmax(double a, double b)
{
    if(a > b)
        return a;
    else
        return b;
};
double fmin(double a, double b)
{
    if(a < b)
        return a;
    else
        return b;
};

fnt_Font *fnt_initFont(const char *filePath, SDL_Renderer *renderer, fnt_float charW, fnt_float lineH, 
        fnt_Charset charset, fnt_uInt16 flags)
{
    /* The exit error codes used by the 'err' label */
    enum EXIT_CODES {
        EXIT_OK                     = 0,
        ERR_ALLOC_FNT_STRUCT        = 1,
        ERR_INIT_FREETYPE           = 2,
        ERR_INIT_FT_FACE            = 3,
        ERR_FT_SET_PIXEL_SIZES      = 4,
        ERR_ALLOC_FNT_GLYPHS        = 5,
        ERR_FT_LOAD_CHAR            = 7,
        ERR_CREATE_CHAR_SURFACE     = 8,
        ERR_CREATE_FINAL_SURFACE    = 9,
        ERR_CREATE_FNT_TEXTURE      = 10,
    };
    int errCode = EXIT_OK;
    int FreeType_errCode = 0;

    char c = 0; /* Used later, but the declaration must be here, as it's needed by the 'err' label */

    fnt_Font *newFont = malloc(sizeof(fnt_Font));
    if(newFont == NULL){
        errCode = ERR_ALLOC_FNT_STRUCT;
        goto err;
    }

    /* Initialize FreeType and load the font */
    FT_Library ft;
    FreeType_errCode = FT_Init_FreeType(&ft);
    if(FreeType_errCode){
        errCode = ERR_INIT_FREETYPE;
        goto err;
    }
    FT_Face face;
    FreeType_errCode = FT_New_Face(ft, filePath, 0, &face);
    if(FreeType_errCode){
        errCode = ERR_INIT_FT_FACE;
        goto err;
    };
    FreeType_errCode = FT_Set_Pixel_Sizes(face, 0, lineH);
    if(FreeType_errCode){
        errCode = ERR_FT_SET_PIXEL_SIZES;
        goto err;
    }

    /* Populate the new font struct with the given/default data */
    newFont->charW = charW;
    newFont->lineHeight = lineH;
    newFont->charset = charset;
    newFont->tabWidth = FNT_DEFAULT_TAB_WIDTH;
    newFont->flags = flags;

    int firstVisibleChar = 0, lastVisibleChar = 0, totalVisibleChars = 0;
    switch ( charset ) {
        default: case FNT_CHARSET_ASCII:
            firstVisibleChar = FNT_ASCII_FIRST_VISIBLE_CHAR;
            lastVisibleChar = FNT_ASCII_LAST_VISIBLE_CHAR;
            totalVisibleChars = FNT_ASCII_TOTAL_VISIBLE_CHARS;
            break;
        case FNT_CHARSET_UTF8:
            // temporary
            firstVisibleChar = FNT_ASCII_FIRST_VISIBLE_CHAR;
            lastVisibleChar = FNT_ASCII_LAST_VISIBLE_CHAR;
            totalVisibleChars = FNT_ASCII_TOTAL_VISIBLE_CHARS;
            break;
    };
    newFont->visibleChars.first = firstVisibleChar;
    newFont->visibleChars.last = lastVisibleChar;
    newFont->visibleChars.total = totalVisibleChars;

    newFont->glyphs = calloc(totalVisibleChars, sizeof(fnt_GlyphData));
    if(newFont->glyphs == NULL){
        errCode = ERR_ALLOC_FNT_GLYPHS;
        goto err;
    }

    /* Prepare the temporary surfaces and the total width and height, 
     * used later to create the final surface with the appropriate dimensions */
    SDL_Surface** tmpSurfaces = malloc(sizeof(SDL_Surface*) * totalVisibleChars);
    int totalWidth = 0, totalHeight = 0;

    for(int i = 0; i < totalVisibleChars; i++){
        fnt_GlyphData *currentGlyph = &newFont->glyphs[i];

        /* char */ c = (char)(i + firstVisibleChar);
        if(FT_Load_Char(face, c, FT_LOAD_RENDER)){
            errCode = ERR_FT_LOAD_CHAR;
            goto err;
        }
        FT_GlyphSlot glyphSlot = face->glyph;

        /* Set the glyph's source rect to based on the metadata from the FT_GlyphSlot struct
         * and the current total width (which also happens to be where our 'pen' is on the X axis)
         */
        currentGlyph->srcRect = (SDL_Rect){
            .x = totalWidth,
            .y = 0,
            .w = glyphSlot->bitmap.width,
            .h = glyphSlot->bitmap.rows,
        };

        /* For some reason FreeType uses 1/64th of a pixel as its metrics unit,
         * so to obtain our wanted result we multiply everything by 64 (shift 6 bits to the right)
         */

        FT_Glyph_Metrics *m = &glyphSlot->metrics;
        if(m->horiAdvance != 0)
            currentGlyph->scaleX = (float)m->width / m->horiAdvance;

        if(lineH != 0)
            currentGlyph->scaleY = (float)m->height / (float)((int)lineH << 6);

        if(m->width != 0)
            currentGlyph->offsetX = ((m->horiBearingX / m->horiAdvance) >> 6) * currentGlyph->scaleX;

        if(m->height != 0)
            currentGlyph->offsetY = 1+0 * currentGlyph->scaleY;

        /* The font's texture will be a long (horizontal) line
         * with all the character placed one after the other,
         * so we increment the total width by the current glyph's width,
         * and only set the height to be the one of the taller character
         */
        totalWidth += currentGlyph->srcRect.w;
        totalHeight = max(totalHeight, glyphSlot->metrics.height >> 6);

        tmpSurfaces[i] = SDL_CreateRGBSurfaceWithFormat(
                0, glyphSlot->bitmap.width, glyphSlot->bitmap.rows, 32, SDL_PIXELFORMAT_RGBA32
            );
        SDL_Surface *currentSurface = tmpSurfaces[i];
        if(currentSurface == NULL){
            errCode = ERR_CREATE_CHAR_SURFACE;
            goto err;
        }

        /* Fill the surface with a transparent black (RGBA 0 0 0 0) background */
        SDL_LockSurface(currentSurface);
        SDL_FillRect(currentSurface, NULL, SDL_MapRGBA(currentSurface->format, 0, 0, 0, 0));

        /* Go through all the pixels one by one, 
         * and map them over onto the current glyph's temp surface */
        for (int y = 0; y < glyphSlot->bitmap.rows; ++y) {
            for (int x = 0; x < glyphSlot->bitmap.width; ++x) {
                Uint8 pixel = glyphSlot->bitmap.buffer[y * glyphSlot->bitmap.width + x];
                Uint32* target_pixel = (Uint32*)((Uint8*)currentSurface->pixels + y * currentSurface->pitch + x * sizeof(Uint32));
                *target_pixel = SDL_MapRGBA(currentSurface->format, pixel, pixel, pixel, pixel);
            }
        }
        SDL_UnlockSurface(currentSurface);
    }

    FT_Done_Face(face);
    FT_Done_FreeType(ft);

    SDL_Surface *finalSurface = SDL_CreateRGBSurfaceWithFormat(
            0, totalWidth, totalHeight, 32, SDL_PIXELFORMAT_RGBA32
        );
    if(finalSurface == NULL){
        errCode = ERR_CREATE_FINAL_SURFACE;
        goto err;
    }
    
    /* Iterate through the temporary surfaces and draw each of them, 
     * one by one, onto the final surface */
    for(int i = 0; i < totalVisibleChars; i++){
        SDL_BlitSurface(tmpSurfaces[i], NULL, finalSurface, &newFont->glyphs[i].srcRect);

        SDL_FreeSurface(tmpSurfaces[i]);
        tmpSurfaces[i] = NULL;
    }
    free(tmpSurfaces);
    tmpSurfaces = NULL;

    /* Now, create the new font's texture from the final temp surface */
    newFont->texture = SDL_CreateTextureFromSurface(renderer, finalSurface);
    if(newFont->texture == NULL){
        errCode = ERR_CREATE_FNT_TEXTURE;
        goto err;
    }
    SDL_FreeSurface(finalSurface);

    SDL_SetTextureBlendMode(newFont->texture, SDL_BLENDMODE_BLEND);
    fnt_setTextColor(newFont, 
        FNT_DEFAULT_TEXT_COLOR.r, 
        FNT_DEFAULT_TEXT_COLOR.g, 
        FNT_DEFAULT_TEXT_COLOR.b, 
        FNT_DEFAULT_TEXT_COLOR.a
    );

    return newFont;

err:
    /* Print the error message BEFORE attempting deallocation of used resources */
    switch(errCode){
        /* We don't care about 'DRY' here, totally not bc im too lazy */
        default: case EXIT_OK:
            fprintf(stderr, 
                "[fnt_initFont]: Something is bad in the code; this situation should never occur. Not freeing any resources allocated in this function!\n"
            );
            break;
    	case ERR_CREATE_FNT_TEXTURE:
	        fprintf(stderr, 
	        	"[fnt_initFont]: Failed to create the font's texture. Reason: %s.\n",
	        	 SDL_GetError()
	        );
			break;
    	case ERR_CREATE_FINAL_SURFACE:
	        fprintf(stderr, 
	        	"[fnt_initFont]: Failed to create the final surface the font's texture. Reason: %s.\n",
	        	 SDL_GetError()
	        );
			break;
    	case ERR_CREATE_CHAR_SURFACE:
	        fprintf(stderr, 
	        	"[fnt_initFont]: Failed to create a surface for the character \'%c\'. Reason: %s.\n",
	        	 c, SDL_GetError()
	        );
			break;
    	case ERR_FT_LOAD_CHAR:
	        fprintf(stderr, 
	        	"[fnt_loadFont]: Failed to load the character \'%c\'. Reason: %s",
	        	 c, FT_Error_String(FreeType_errCode)
	        );
			break;
    	case ERR_ALLOC_FNT_GLYPHS:
	        fprintf(stderr, "[fnt_loadFont]: Failed to allocate memory for the new font's glyphs.\n");
			break;
        case ERR_FT_SET_PIXEL_SIZES:
	        fprintf(stderr, 
                "[fnt_loadFont]: Failed set pixel sizes for the new font. Reason:%s.\n",
                FT_Error_String(FreeType_errCode)
            );
            break;
    	case ERR_INIT_FT_FACE:
	        fprintf(stderr, 
	        	"[fnt_loadFont]: Failed to create a FreeType face. Reason: %s.\n",
	        	 FT_Error_String(FreeType_errCode)
	        );
			break;
    	case ERR_INIT_FREETYPE:
	        fprintf(stderr, 
	        	"[fnt_loadFont]: Failed to initialize FreeType. Reason: %s.\n",
	        	 FT_Error_String(FreeType_errCode)
	        );
			break;
    	case ERR_ALLOC_FNT_STRUCT:
	        fprintf(stderr, "[fnt_loadFont]: Failed to allocate memory for the new font struct.\n");
			break;
    }
    
    /* Depending on the stage of the initialization at which the error occured,
     * free the resources that might have been allocated up to that point 
     *
     * (We are assuming tha the error codes are ordered appropriately in the EXIT_CODES enum) 
    */
    if(errCode >= ERR_CREATE_FNT_TEXTURE){
        SDL_FreeSurface(finalSurface);
        finalSurface = NULL;
    }
    if(errCode >= ERR_CREATE_FINAL_SURFACE){
        for(int i = 0; i < totalVisibleChars; i++){
            SDL_FreeSurface(tmpSurfaces[i]);
            tmpSurfaces[i] = NULL;
        }
    }
    if(errCode >= ERR_FT_LOAD_CHAR){
        free(newFont->glyphs);
        newFont->glyphs = NULL;
    }
    if(errCode >= ERR_FT_SET_PIXEL_SIZES){
        FT_Done_Face(face);
    }
    if(errCode >= ERR_INIT_FT_FACE){
        FT_Done_FreeType(ft);
    }
    if(errCode >= ERR_INIT_FREETYPE){
        free(newFont);
        newFont = NULL;
    }

    return NULL;
}

void fnt_drawText(fnt_Font *fnt, SDL_Renderer *renderer, fnt_Vector2D *pos, const char *fmt, ...)
{
    va_list vArgs;
    char str[FNT_TEXT_BUFFER_SIZE];

    va_start(vArgs, fmt);
    vsnprintf(str, FNT_TEXT_BUFFER_SIZE - 1, fmt, vArgs);
    va_end(vArgs);
    str[FNT_TEXT_BUFFER_SIZE - 1] = '\0';

    /* penX and penY are relative to the given baseline coorinates */
    int penX = 0, penY = 0;
    int maxPenX = 0;

    const char *c_ptr = str;
    int i;
    
    while(*c_ptr){
        switch(*c_ptr){
            default:
                i = (int)(*c_ptr) - fnt->visibleChars.first;
                if(i < fnt->visibleChars.total && i >= 0){
                    fnt_GlyphData *currentGlyph = &fnt->glyphs[i];

                    SDL_Rect glyphDestRect = { 
                        .x = pos->x + penX + (currentGlyph->offsetX * fnt->charW),
                        .y = pos->y + penY + (0*currentGlyph->offsetY * fnt->lineHeight),
                        .w = (int)(fnt->charW * currentGlyph->scaleX),
                        .h = (int)(fnt->lineHeight * currentGlyph->scaleY),
                    };
                    SDL_RenderCopy(renderer, fnt->texture, &fnt->glyphs[i].srcRect, &glyphDestRect);
                    if(fnt->flags & FNT_FLAG_DISPLAY_GLYPH_RECTS){
                        SDL_SetRenderDrawColor(renderer, 255, 0, 0, 255);
                        SDL_RenderDrawRect(renderer, &glyphDestRect);
                    }

                    SDL_Rect charDestRect = (SDL_Rect){
                        .x = pos->x + penX, 
                        .y = pos->y + penY,
                        .w = fnt->charW,
                        .h = fnt->lineHeight
                    };
                    if(fnt->flags & FNT_FLAG_DISPLAY_CHAR_RECTS){
                        SDL_SetRenderDrawColor(renderer, 255, 255, 0, 255);
                        SDL_RenderDrawRect(renderer, &charDestRect);
                    }

                    penX += fnt->charW;
                }
                break;
            case ' ':
                penX += fnt->charW;
                break;
            case '\r': /* Carriage return */
                penX = 0;
                break;
            case '\t':
                /* Advance to the next multiple of tabWidth (times the character width) */
                if(fnt->charW != 0)
                    penX += (fnt->tabWidth -((int)(penX / fnt->charW) % fnt->tabWidth)) * fnt->charW;
                break;
            case '\n':
                maxPenX = max(penX, maxPenX);
                penX = 0;
                penY += fnt->lineHeight;
                break;
            case '\f': /* Form feed character, just treat it as two newline characters */
                penX = 0;
                penY += fnt->lineHeight * 2;
                break;
            case '\v': /* Vertical tab (newline + tab) */
                penY += fnt->lineHeight;
                penX = fnt->tabWidth * fnt->charW;
                break;
        }
        c_ptr++;
    }
    if(fnt->flags & FNT_FLAG_DISPLAT_TEXT_RECTS){
        SDL_Rect textDestRect = {
            .x = pos->x,
            .y = pos->y,
            .w = max(penX, maxPenX),
            .h = penY + fnt->lineHeight,
        };
        SDL_SetRenderDrawColor(renderer, 0, 255, 0, 255);
        SDL_RenderDrawRect(renderer, &textDestRect);
    }
}

void fnt_destroyFont(fnt_Font *fnt)
{
    free(fnt->glyphs);
    fnt->glyphs = NULL;

    SDL_DestroyTexture(fnt->texture);

    free(fnt);
    fnt = NULL;
}

void fnt_setTextColor(fnt_Font *fnt, uint8_t r, uint8_t g, uint8_t b, uint8_t a)
{
    SDL_SetTextureColorMod(fnt->texture, r, g, b);
    SDL_SetTextureAlphaMod(fnt->texture, a);
}
